File: _archive/freee-sync-ts/src/sync.ts
/**
 * Sync orchestration
 * Fetches data from freee and converts to Beancount format
 * With SQLite-based duplicate prevention
 */

import * as fs from 'fs';
import * as path from 'path';
import { FreeeClient } from './freee/client';
import { Deal, Journal } from './freee/types';
import { AccountMapper } from './converter/mapper';
import { BeancountConverter } from './converter/converter';
import { DatabaseConnection, createConnection } from './db/connection';
import { SyncHistory } from './db/sync-history';

export interface SyncOptions {
  dateFrom: string;
  dateTo: string;
  beancountDir: string;
  dryRun?: boolean;
}

export interface SyncResult {
  dealsCount: number;
  journalsCount: number;
  filesWritten: string[];
}

export class SyncOrchestrator {
  private client: FreeeClient;
  private mapper: AccountMapper;
  private converter: BeancountConverter;
  private dbConnection: DatabaseConnection;
  private syncHistory: SyncHistory;

  constructor(
    client: FreeeClient,
    mapper: AccountMapper,
    dbConnection?: DatabaseConnection
  ) {
    this.client = client;
    this.mapper = mapper;
    this.converter = new BeancountConverter(mapper);
    this.dbConnection = dbConnection || createConnection();
    this.syncHistory = new SyncHistory(this.dbConnection);
  }

  /**
   * Perform sync operation with duplicate prevention
   */
  async sync(options: SyncOptions): Promise<SyncResult> {
    console.log(`Syncing from ${options.dateFrom} to ${options.dateTo}`);

    // Fetch deals from freee
    console.log('Fetching deals from freee...');
    const allDeals = await this.client.fetchAllDeals(options.dateFrom, options.dateTo);
    console.log(`Fetched ${allDeals.length} deals`);

    // Fetch journals from freee
    console.log('Fetching journals from freee...');
    const allJournals = await this.client.fetchAllJournals(options.dateFrom, options.dateTo);
    console.log(`Fetched ${allJournals.length} journals`);

    // Filter out already synced items
    console.log('Checking for already synced items...');
    const syncedDealIds = await this.syncHistory.getSyncedIds('deal');
    const syncedJournalIds = await this.syncHistory.getSyncedIds('journal');

    const newDeals = allDeals.filter((d) => !syncedDealIds.includes(d.id));
    const newJournals = allJournals.filter((j) => !syncedJournalIds.includes(j.id));

    console.log(
      `New items: ${newDeals.length} deals, ${newJournals.length} journals ` +
      `(skipped: ${allDeals.length - newDeals.length} deals, ${allJournals.length - newJournals.length} journals)`
    );

    if (newDeals.length === 0 && newJournals.length === 0) {
      console.log('No new items to sync');
      return {
        dealsCount: 0,
        journalsCount: 0,
        filesWritten: [],
      };
    }

    // Group by month
    const dealsByMonth = this.groupByMonth(newDeals, (d) => d.issue_date);
    const journalsByMonth = this.groupByMonth(newJournals, (j) => j.issue_date);

    // Get all unique months
    const allMonths = new Set([
      ...Object.keys(dealsByMonth),
      ...Object.keys(journalsByMonth),
    ]);

    const filesWritten: string[] = [];

    // Process each month
    for (const monthKey of Array.from(allMonths).sort()) {
      const monthDeals = dealsByMonth[monthKey] || [];
      const monthJournals = journalsByMonth[monthKey] || [];

      const filePath = this.getMonthFilePath(monthKey, options.beancountDir);

      if (!options.dryRun) {
        // Append to existing file or create new
        this.appendBeancountTransactions(filePath, monthDeals, monthJournals);
        filesWritten.push(filePath);
        console.log(`Updated ${filePath} (${monthDeals.length} deals, ${monthJournals.length} journals)`);

        // Record sync history
        for (const deal of monthDeals) {
          await this.syncHistory.recordSync({
            sync_type: 'deal',
            freee_id: deal.id,
            issue_date: deal.issue_date,
            amount: deal.amount,
            beancount_file: filePath,
          });
        }

        for (const journal of monthJournals) {
          await this.syncHistory.recordSync({
            sync_type: 'journal',
            freee_id: journal.id,
            issue_date: journal.issue_date,
            amount: journal.details[0]?.amount || 0,
            beancount_file: filePath,
          });
        }
      } else {
        console.log(`[DRY RUN] Would append to ${filePath}`);
        for (const deal of monthDeals) {
          const txn = this.converter.convertDeal(deal);
          console.log(this.converter.formatTransaction(txn));
        }
        for (const journal of monthJournals) {
          const txn = this.converter.convertJournal(journal);
          console.log(this.converter.formatTransaction(txn));
        }
      }
    }

    // Display statistics
    const stats = await this.syncHistory.getStats();
    console.log('\n=== Sync Statistics ===');
    console.log(`Total synced deals: ${stats.total_deals}`);
    console.log(`Total synced journals: ${stats.total_journals}`);
    console.log(`Total documents: ${stats.total_documents}`);
    console.log(`Last sync: ${stats.last_sync}`);

    return {
      dealsCount: newDeals.length,
      journalsCount: newJournals.length,
      filesWritten,
    };
  }

  /**
   * Group items by month (YYYY-MM)
   */
  private groupByMonth<T>(
    items: T[],
    getDate: (item: T) => string
  ): Record<string, T[]> {
    const groups: Record<string, T[]> = {};

    for (const item of items) {
      const date = getDate(item);
      const monthKey = date.substring(0, 7); // YYYY-MM

      if (!groups[monthKey]) {
        groups[monthKey] = [];
      }

      groups[monthKey].push(item);
    }

    return groups;
  }

  /**
   * Generate Beancount file content for a month
   */
  private generateBeancountFile(deals: Deal[], journals: Journal[]): string {
    let content = '';

    // Header comment
    if (deals.length > 0 || journals.length > 0) {
      const month = deals.length > 0 ? deals[0].issue_date.substring(0, 7) : journals[0].issue_date.substring(0, 7);
      content += `; freee sync - ${month}\n`;
      content += `; Generated at ${new Date().toISOString()}\n`;
      content += `; Deals: ${deals.length}, Journals: ${journals.length}\n`;
      content += '\n';
    }

    // Convert deals
    for (const deal of deals) {
      const txn = this.converter.convertDeal(deal);
      content += this.converter.formatTransaction(txn);
      content += '\n';
    }

    // Convert journals
    for (const journal of journals) {
      const txn = this.converter.convertJournal(journal);
      content += this.converter.formatTransaction(txn);
      content += '\n';
    }

    return content;
  }

  /**
   * Get file path for a month
   */
  private getMonthFilePath(monthKey: string, beancountDir: string): string {
    const [year, month] = monthKey.split('-');
    const yearDir = path.join(beancountDir, year);
    const fileName = `${year}-${month}.beancount`;

    return path.join(yearDir, fileName);
  }

  /**
   * Write Beancount file
   */
  private writeBeancountFile(filePath: string, content: string): void {
    const dir = path.dirname(filePath);

    // Ensure directory exists
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    fs.writeFileSync(filePath, content, 'utf-8');
  }

  /**
   * Append transactions to Beancount file
   * Creates new file if it doesn't exist
   */
  private appendBeancountTransactions(
    filePath: string,
    deals: Deal[],
    journals: Journal[]
  ): void {
    const dir = path.dirname(filePath);

    // Ensure directory exists
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    // Check if file exists
    const fileExists = fs.existsSync(filePath);

    if (!fileExists) {
      // Create new file with header
      const monthKey = path.basename(filePath, '.beancount');
      let content = `; freee sync - ${monthKey}\n`;
      content += `; Generated at ${new Date().toISOString()}\n\n`;
      fs.writeFileSync(filePath, content, 'utf-8');
    }

    // Append transactions
    let content = '';

    for (const deal of deals) {
      const txn = this.converter.convertDeal(deal);
      content += this.converter.formatTransaction(txn);
      content += '\n';
    }

    for (const journal of journals) {
      const txn = this.converter.convertJournal(journal);
      content += this.converter.formatTransaction(txn);
      content += '\n';
    }

    // Append to file
    fs.appendFileSync(filePath, content, 'utf-8');
  }

  /**
   * Close database connection
   * Should be called when done with sync operations
   */
  close(): void {
    this.dbConnection.close();
  }
}

File: cmd/freee-sync/cmd/sync.go
package cmd

import (
	"fmt"
	"log/slog"
	"path/filepath"
	"time"

	"github.com/shunichi-ikebuchi/accounting-system/pkg/beancount"
	"github.com/shunichi-ikebuchi/accounting-system/pkg/config"
	"github.com/shunichi-ikebuchi/accounting-system/pkg/converter"
	"github.com/shunichi-ikebuchi/accounting-system/pkg/db"
	"github.com/shunichi-ikebuchi/accounting-system/pkg/freee"
	"github.com/shunichi-ikebuchi/accounting-system/pkg/pathutil"
	"github.com/spf13/cobra"
)

var (
	dateFrom string
	dateTo   string
	dryRun   bool
)

// syncCmd represents the sync command.
var syncCmd = &cobra.Command{
	Use:   "sync",
	Short: "Sync freee transactions to Beancount",
	Long: `Sync transactions from freee Accounting API to Beancount files.

This command:
1. Fetches deals and journals from freee API
2. Filters out already synced items
3. Converts them to Beancount format
4. Appends to monthly Beancount files
5. Records sync history in SQLite

Example:
  freee-sync sync --from 2024-01-01 --to 2024-01-31
  freee-sync sync --from 2024-01-01 --to 2024-01-31 --dry-run`,
	Run: runSync,
}

func init() {
	// Flags
	syncCmd.Flags().StringVar(&dateFrom, "from", "", "Start date (YYYY-MM-DD) (required)")
	syncCmd.Flags().StringVar(&dateTo, "to", "", "End date (YYYY-MM-DD) (required)")
	syncCmd.Flags().BoolVar(&dryRun, "dry-run", false, "Dry run mode (no file writes)")

	syncCmd.MarkFlagRequired("from")
	syncCmd.MarkFlagRequired("to")
}

func runSync(cmd *cobra.Command, args []string) {
	slog.Info("Starting sync", "from", dateFrom, "to", dateTo, "dry_run", dryRun)

	// Load configuration
	cfg, err := config.Load(getConfigFile())
	exitOnError(err, "failed to load configuration")

	// Validate required fields
	if err := cfg.Validate(
		[]string{"freee", "apiUrl"},
		[]string{"freee", "accessToken"},
		[]string{"freee", "companyId"},
		[]string{"beancount", "root"},
	); err != nil {
		exitOnError(err, "invalid configuration")
	}

	// Initialize components
	pathResolver := pathutil.New(pathutil.Config{
		BeancountRoot:  cfg.Beancount.Root,
		DatabasePath:   cfg.Beancount.DBPath,
		AttachmentsDir: cfg.Beancount.AttachmentsDir,
	})

	// Open database
	dbPath := pathResolver.GetDatabasePath()
	slog.Debug("Opening database", "path", dbPath)
	conn, err := db.Open(dbPath)
	exitOnError(err, "failed to open database")
	defer conn.Close()

	syncHistory := db.NewSyncHistory(conn)

	// Initialize freee API client
	freeeClient := freee.NewClient(freee.ClientConfig{
		APIURL:      cfg.Freee.APIURL,
		AccessToken: cfg.Freee.AccessToken,
		CompanyID:   cfg.Freee.CompanyID,
		Timeout:     30 * time.Second,
	})

	// Initialize account mapper
	mappingFilePath := filepath.Join("config", "account-mapping.yaml")
	mapper, err := converter.NewMapper(mappingFilePath)
	exitOnError(err, "failed to load account mapping")

	// Initialize converter
	cvtr := converter.NewConverter(mapper, "JPY")

	// Initialize Beancount repository
	beancountRepo := beancount.NewFileSystemRepository(pathResolver)

	// Fetch deals from freee
	slog.Info("Fetching deals from freee", "from", dateFrom, "to", dateTo)
	allDeals, err := freeeClient.FetchAllDeals(dateFrom, dateTo)
	exitOnError(err, "failed to fetch deals")
	slog.Info("Fetched deals", "count", len(allDeals))

	// Fetch journals from freee
	slog.Info("Fetching journals from freee", "from", dateFrom, "to", dateTo)
	allJournals, err := freeeClient.FetchAllJournals(dateFrom, dateTo)
	exitOnError(err, "failed to fetch journals")
	slog.Info("Fetched journals", "count", len(allJournals))

	// Filter out already synced items
	slog.Info("Checking for already synced items")
	syncedDealIDs, err := syncHistory.GetSyncedIDs(db.SyncTypeDeal)
	exitOnError(err, "failed to get synced deal IDs")

	syncedJournalIDs, err := syncHistory.GetSyncedIDs(db.SyncTypeJournal)
	exitOnError(err, "failed to get synced journal IDs")

	newDeals := filterDeals(allDeals, syncedDealIDs)
	newJournals := filterJournals(allJournals, syncedJournalIDs)

	slog.Info("New items to sync",
		"new_deals", len(newDeals),
		"new_journals", len(newJournals),
		"skipped_deals", len(allDeals)-len(newDeals),
		"skipped_journals", len(allJournals)-len(newJournals),
	)

	if len(newDeals) == 0 && len(newJournals) == 0 {
		fmt.Println("No new items to sync")
		return
	}

	// Group by month
	dealsByMonth := groupDealsByMonth(newDeals)
	journalsByMonth := groupJournalsByMonth(newJournals)

	// Get all unique months
	allMonths := getAllMonths(dealsByMonth, journalsByMonth)

	filesWritten := []string{}

	// Process each month
	for _, monthKey := range allMonths {
		monthDeals := dealsByMonth[monthKey]
		monthJournals := journalsByMonth[monthKey]

		filePath, err := pathResolver.GetMonthFilePath(monthKey)
		if err != nil {
			slog.Error("Failed to get month file path", "month", monthKey, "error", err)
			continue
		}

		if !dryRun {
			// Ensure month file exists
			if err := beancountRepo.EnsureMonthFile(monthKey); err != nil {
				slog.Error("Failed to ensure month file", "month", monthKey, "error", err)
				continue
			}

			// Append transactions
			for _, deal := range monthDeals {
				txn := cvtr.ConvertDeal(deal)
				formatted := cvtr.FormatTransaction(txn)

				if err := beancountRepo.AppendTransaction(monthKey, formatted); err != nil {
					slog.Error("Failed to append deal", "deal_id", deal.ID, "error", err)
					continue
				}

				// Record sync history
				if err := syncHistory.RecordSync(db.SyncRecord{
					SyncType:      db.SyncTypeDeal,
					FreeeID:       deal.ID,
					IssueDate:     deal.IssueDate,
					Amount:        deal.Amount,
					BeancountFile: filePath,
				}); err != nil {
					slog.Error("Failed to record sync", "deal_id", deal.ID, "error", err)
				}
			}

			for _, journal := range monthJournals {
				txn := cvtr.ConvertJournal(journal)
				formatted := cvtr.FormatTransaction(txn)

				if err := beancountRepo.AppendTransaction(monthKey, formatted); err != nil {
					slog.Error("Failed to append journal", "journal_id", journal.ID, "error", err)
					continue
				}

				amount := int64(0)
				if len(journal.Details) > 0 {
					amount = journal.Details[0].Amount
				}

				// Record sync history
				if err := syncHistory.RecordSync(db.SyncRecord{
					SyncType:      db.SyncTypeJournal,
					FreeeID:       journal.ID,
					IssueDate:     journal.IssueDate,
					Amount:        amount,
					BeancountFile: filePath,
				}); err != nil {
					slog.Error("Failed to record sync", "journal_id", journal.ID, "error", err)
				}
			}

			filesWritten = append(filesWritten, filePath)
			slog.Info("Updated file",
				"path", filePath,
				"deals", len(monthDeals),
				"journals", len(monthJournals),
			)
		} else {
			// Dry run: print transactions
			fmt.Printf("[DRY RUN] Would append to %s\n", filePath)
			for _, deal := range monthDeals {
				txn := cvtr.ConvertDeal(deal)
				fmt.Println(cvtr.FormatTransaction(txn))
			}
			for _, journal := range monthJournals {
				txn := cvtr.ConvertJournal(journal)
				fmt.Println(cvtr.FormatTransaction(txn))
			}
		}
	}

	// Display final statistics
	if !dryRun {
		stats, err := syncHistory.GetStats()
		if err == nil {
			fmt.Println("\n=== Sync Statistics ===")
			fmt.Printf("Total synced deals:    %d\n", stats.TotalDeals)
			fmt.Printf("Total synced journals: %d\n", stats.TotalJournals)
			fmt.Printf("Total documents:       %d\n", stats.TotalDocuments)
			if stats.LastSync.Valid {
				fmt.Printf("Last sync:             %s\n", stats.LastSync.String)
			}
			fmt.Println()
		}
	}

	slog.Info("Sync completed",
		"new_deals", len(newDeals),
		"new_journals", len(newJournals),
		"files_written", len(filesWritten),
	)
}

// Helper functions

func filterDeals(deals []freee.Deal, syncedIDs []int64) []freee.Deal {
	syncedIDMap := make(map[int64]bool)
	for _, id := range syncedIDs {
		syncedIDMap[id] = true
	}

	var result []freee.Deal
	for _, deal := range deals {
		if !syncedIDMap[deal.ID] {
			result = append(result, deal)
		}
	}
	return result
}

func filterJournals(journals []freee.Journal, syncedIDs []int64) []freee.Journal {
	syncedIDMap := make(map[int64]bool)
	for _, id := range syncedIDs {
		syncedIDMap[id] = true
	}

	var result []freee.Journal
	for _, journal := range journals {
		if !syncedIDMap[journal.ID] {
			result = append(result, journal)
		}
	}
	return result
}

func groupDealsByMonth(deals []freee.Deal) map[string][]freee.Deal {
	groups := make(map[string][]freee.Deal)
	for _, deal := range deals {
		monthKey := deal.IssueDate[:7] // YYYY-MM
		groups[monthKey] = append(groups[monthKey], deal)
	}
	return groups
}

func groupJournalsByMonth(journals []freee.Journal) map[string][]freee.Journal {
	groups := make(map[string][]freee.Journal)
	for _, journal := range journals {
		monthKey := journal.IssueDate[:7] // YYYY-MM
		groups[monthKey] = append(groups[monthKey], journal)
	}
	return groups
}

func getAllMonths(dealGroups map[string][]freee.Deal, journalGroups map[string][]freee.Journal) []string {
	monthsMap := make(map[string]bool)
	for month := range dealGroups {
		monthsMap[month] = true
	}
	for month := range journalGroups {
		monthsMap[month] = true
	}

	months := []string{}
	for month := range monthsMap {
		months = append(months, month)
	}

	// Sort months (simple string sort works for YYYY-MM format)
	for i := 0; i < len(months); i++ {
		for j := i + 1; j < len(months); j++ {
			if months[i] > months[j] {
				months[i], months[j] = months[j], months[i]
			}
		}
	}

	return months
}
